data_scrapers/san_francisco.py

```diff
@@ -0,0 +1,268 @@
+    out["update_time"] = sorted(update_times)[0]  # get earliest update time
+    out["meta_from_source"] = meta_from_source
+    out["meta_from_baypd"] = "SF county only reports tests with positive or negative results, excluding pending tests. The following datapoints are not directly reported, and were calculated by BayPD using available data: cumulative cases, cumulative deaths, cumulative positive tests, cumulative negative tests, cumulative total tests."
+
+    # get timeseries and demographic totals
+    out["series"] = session.get_timeseries()
+    demo_totals = session.get_demographics()
+    out.update(demo_totals)
+
+    return out
+
+
+class SanFranciscoApi(SocrataApi):
+    """
+    Inherits from utils.SocratApi. Creates a session to query the resource ids.
+    Re-keys and re-structures data to fit our data model.
+    """
+
+    def __init__(self):
+        super().__init__('https://data.sfgov.org/')
+        self.resource_ids = {'cases_deaths_transmission': 'tvq9-ec9w', 'age_gender': 'sunc-2t3k',
+                             'race_eth': 'vqqm-nsqg', 'tests': 'nfpa-mg4g'}
+
+
+    def request(self, resource_id, **kwargs):
+        return super().request(f'{self.resource_url}{resource_id}', **kwargs)
+
+    def get_notes(self) -> str:
+        """
+        Get 'description' field of metadata for all resources. Collect into one string,
+        separated by 2 newlines.
+        """
+        meta_from_source = ''
+        for v in self.resource_ids.values():
+            url = f"{self.metadata_url}{v}.json"
+            data = super().request(url)
+            meta_from_source += data["description"] + '\n\n'
+        return meta_from_source
+
+    def get_update_times(self) -> List:
+        """
+        Return a list of update times for all resources.
+        """
+        update_times = []
+        for v in self.resource_ids.values():
+            url = f"{self.metadata_url}{v}.json"
+            data = super().request(url)
+            update_times.append(data["dataUpdatedAt"])
+        return update_times
+
+    def get_demographics(self) -> Dict:
+        """
+        Fetch cases by age, gender, race_eth. Fetch cases by transmission category
+        Returns the dictionary value for {"cases_totals": {}, "death_totals":{}}.
+        Note that SF does not provide death totals, so these datapoints will be -1.
+        To crete a DataFrame from the dictionary, run 'pd.DataFrame(get_demographics())'
+        """
+        # copy dictionary structure of global 'out' dictionary to local variable
+        demo_totals: Dict[str,Dict] = { "case_totals": dict(), "death_totals": dict() }
+        demo_totals["case_totals"]["gender"] = self.get_gender_table()
+        demo_totals["case_totals"]["age_group"] = self.get_age_table()
+        demo_totals["case_totals"]["transmission_cat"] = self.get_transmission_table()
+        demo_totals["case_totals"]["race_eth"] = self.get_race_eth_table()
+        return demo_totals
```

Comment:
I’ve lost track in this refactor: did we previously fill in `death_totals` for SF? It’s missing now.

(If no, it might good to add a `TODO` comment in the source calling out that this field needs work to fill in.)

Reply:
We didn't have death_totals (and still don't). Jason Lally says it should be up this week. I'll add a `TODO` to fill in.